### ISSUE TITLE
Monitor panels forgotten between connections

### ISSUE THREAD

**Issue by rmwiseman** [linuxmint/cinnamon#12467]
### Distribution

22

### Package version

6.2.9

### Frequency

Quite often

### Bug description

When I connect to a monitor (I have one at work and one at home) I like to have a panel on the monitor as well as on my laptop screen to keep the window list on the same monitor as the windows.  I add a panel and it works fine.  But then, much of time time, when I come back another day and plug the monitor in again, the panel I added is gone.  Previously, this hasn't been the case, and each monitor's configuration has been remembered.

Also, when I add the new panel, the icons of the windows are huge and are consequently cropped in the panel.  But if I edit the panel settings and change the panel height, it fixes it.  The panel height starts at 25; I move the slider to 26 and then back to 25 and the icons' sizes are correct.

### Steps to reproduce

As far as I can tell, this happens between monitors, so if I'm using the laptop in the office at work, and then come home and plug it in, that's when it tends to happen -- when I've been using a different monitor.  Or vice versa: home to work.  I use USB-C at home and either USB-C or HDMI at work (to the same monitor) if that makes a difference.

### Expected behavior

Cinnamon remembers the monitor and displays the panel as previously configured.

**Comment by rmwiseman** [comment]
And something else I noticed today…

Even though I'd not connected any other monitors since I was last in the office, when I arrived today and  plugged in the first of my two monitors, its panel had disappeared again.  I then connected my second external monitor and that, too, had no panel.

However — and this came as something of a surprise — when I re-added the panel to the first external monitor, the second monitor's panel appeared at exactly the same moment.  Both had the applets I had previously added.  Unfortunately, the window list applet in both of them had very large, cropped icons, and as I indicated previously, I had to change each panel's settings to correct the icon size (specifically, I had to drag the panel height away from 25 and then back to 25).

**Comment by kogans** [comment]
I'm experiencing the same bug - I created issue #12641 for it - it should probably be merged into here.
The missing panel is definitely there - you can see its' properties in the "panel settings" - but it is not visible.

**Comment by rmwiseman** [comment]
_Update following new version of Mint and Cinnamon…_

Mint 22.1
Cinnamon 6.4.6

The issues persist, but the incorrect applet height issue has changed slightly in how it manifests: previously, the icons were cropped, but now they are visible outside the panel:

![Image](https://github.com/user-attachments/assets/cee41206-e1d0-48fd-9b2e-012d42d5611b)

As before, changing the panel height in the panel settings fixes the problem (dragging the slider away and back to where it was).

**Comment by rmwiseman** [comment]
In case anyone else has the same problem as me, where the icons are incorrectly sized, I have written a simple script to fix them.  It works by increasing the height of all panels, pausing briefly, and then reverting them.  (The panel sizes in dconf are correct, but the icons don't always obey their parent panel's size.)  This fix is a programmatic version of editing the panel settings described above.

Note that the script does rely on Perl being installed, but I think it should be by default.

```bash
#!/usr/bin/env bash

# This script can be run when a new monitor has been plugged in and Cinnamon
# has got confused with icon sizes in panels.  If the panels are missing,
# restart Cinnamon (press Ctrl+Alt+Esc).  If the icons now overlap the panel
# (or are cropped, as with previous versions), then this script should sort out
# the icons' sizes.

path=/org/cinnamon/panels-height
correct=$( dconf read $path )
bigger=$( perl -pe 's/:(\d+)/":".($1+1)/ge' <<<$correct )

dconf write $path "$bigger"
sleep 0.5
dconf write $path "$correct"
```

**Comment by theluckyfellow** [comment]
Hit this too. I traced the root cause and implemented + tested a fix on Cinnamon 6.6.7 (Mint 22.3) — sharing in case it helps move this forward.

**TL;DR:** a panel whose monitor is absent **at the moment Cinnamon loads panels** (boot/login/Cinnamon-restart while undocked, or an external that re-enumerates on a different connector during a dock change) is dropped and **never re-created when the monitor reconnects**, even though there's recovery code that's supposed to do exactly that. Only a manual Cinnamon restart brings it back. If it was the only panel, you're left with no panel at all.

### Environment
- Linux Mint 22.3 "Zena", **Cinnamon 6.6.7**, Muffin 6.6.3, X11
- Kernel 6.17, AMD `amdgpu`
- Laptop (`eDP`) + external over a dock. Panel was pinned to the external: `org.cinnamon panels-enabled = ['1:1:right']`. After undocking, the external came back on a **different DisplayPort connector** (`DisplayPort-7` in `~/.config/cinnamon-monitors.xml` → `DisplayPort-6` in `xrandr`), and the panel was gone from both screens. `~/.xsession-errors` showed at startup:
  ```
  JS LOG: [LookingGlass/info] Ignoring panel definition for nonexistent monitor: 1:1:right
  ```

### Deterministic repro (single display, no second monitor needed)
1. `gsettings get org.cinnamon panels-enabled`  *(save it)*
2. `gsettings set org.cinnamon panels-enabled "['1:1:bottom']"`  *(monitor index 1 doesn't exist)*
3. Restart Cinnamon (**Ctrl+Alt+Esc**)
   → no panel is drawn; log shows `Ignoring panel definition for nonexistent monitor: 1:1:bottom`, and there is **no fallback** to monitor 0.
4. Restore: `gsettings set org.cinnamon panels-enabled "['1:0:bottom']"` then Ctrl+Alt+Esc.

To see the non-recovery in the real case: put a panel on an external, restart Cinnamon while it's disconnected, then reconnect it — the panel does not return without a manual Cinnamon restart.

### Root cause (`js/ui/panel.js`, 6.6.7)
1. `_fullPanelLoad()` skips absent-monitor panels **before** they're loaded (~L446):
   ```js
   if (elements[PanelDefElement.MONITOR] >= monitorCount) {
       // Ignore, but don't remove. Less monitors can be a temporary condition.
       global.log("Ignoring panel definition for nonexistent monitor: " + panel_defs[i]);
       continue;                       // never added to good_defs → _loadPanel() never called
   }
   ```
2. `_loadPanel()` is the only place (in this path) that records a panel's meta — and it does so at L889 **before** bailing on a missing monitor at L891:
   ```js
   metaList[ID] = [monitorIndex, panelPosition];                 // 889 (meta recorded)
   if (monitorIndex < 0 || monitorIndex >= this.monitorCount) {  // 891
       global.log("Monitor " + monitorIndex + " not found. Not creating panel");
       return null;                                              // 893
   }
   ```
   So a panel that *reaches* `_loadPanel()` gets a `panelsMeta` entry even with no monitor — but the one skipped in step 1 never reaches it, so it gets **no meta**.
3. The reconnect-recovery in `_onMonitorsChanged()` depends on that meta (~L1126):
   ```js
   if (!this.panelsMeta[i]) { continue; }          // skips meta-less panels
   if (!this.panels[i]) {  // meta but no panel: "the monitor may just have been reconnected"
       if (this.panelsMeta[i][0] < this.monitorCount) {
           let panel = this._loadPanel(i, this.panelsMeta[i][0], this.panelsMeta[i][1], drawcorner);
           if (panel) AppletManager.loadAppletsOnPanel(panel);
       }
   }
   ```
   For a panel dropped at startup, `panelsMeta[i]` is undefined → L1126 `continue` skips it → the recovery branch can never run.

**Net:** monitor removed *after* a panel exists → meta exists → recovers on reconnect (works). Monitor absent *at Cinnamon startup* → no meta → never recovers (this bug).

### A second, related gap (live hot-undock)
Even when a panel *does* have meta, `_onMonitorsChanged()` destroys the panel of a monitor that disappears (the `this.panelsMeta[i][0] >= this.monitorCount` branch) **with no fallback to the primary**. So hot-undocking — not just a restart — can leave you with **zero panels** when your only panel was on the external display. The same "never leave the user with no panel → place it on the primary monitor" safety net is needed in `_onMonitorsChanged()`, not only at load time.

### Fix (implemented and tested on 6.6.7)
Record the meta for a deferred panel so the existing recovery path becomes reachable — replace the L446 block:
```js
if (elements[PanelDefElement.MONITOR] >= monitorCount) {
    // Monitor absent right now (e.g. undocked / not yet enumerated). Don't create
    // the panel, but record its meta so _onMonitorsChanged() can re-create it on
    // reconnect — otherwise it has no panelsMeta entry and recovery never fires.
    global.log("Deferring panel for absent monitor: " + panel_defs[i]);
    this.panelsMeta[parseInt(elements[PanelDefElement.ID])] =
        [parseInt(elements[PanelDefElement.MONITOR]),
         getPanelLocFromName(elements[PanelDefElement.POSITION])];
    continue;
}
```
Optional safety net (never leave the user with zero panels) — after the load loops in `_fullPanelLoad()`:
```js
if (this.panelCount === 0 && panel_defs.length > 0) {
    let prim = global.display.get_primary_monitor();
    let def  = panel_defs[0].split(":");
    let p = this._loadPanel(parseInt(def[PanelDefElement.ID]), prim,
                            getPanelLocFromName(def[PanelDefElement.POSITION]), [true, true]);
    if (p) AppletManager.loadAppletsOnPanel(p);
}
```

(The same zero-panel→primary fallback should also be added at the end of `_onMonitorsChanged()` to cover the live hot-undock case described above.)

The connector renaming that triggers this (`DisplayPort-7` ↔ `DisplayPort-6` for the same physical monitor) is kernel/GPU/dock behaviour, not Cinnamon's — the changes above are about Cinnamon recovering gracefully. I've applied the meta-fix plus a relocate-to-a-present-monitor step on monitor changes on my own 6.6.7 install, and confirmed — via both simulated (`xrandr --output <ext> --off/--auto`) and real dock/undock cycles — that the panel is no longer lost: undocking keeps a panel on the laptop, and reconnecting restores it. Syntax-checked against the cjs/SpiderMonkey engine before installing. This is on one laptop+dock config (single panel), so it'd want testing on multi-panel/multi-monitor setups, but I'm happy to open a PR if a maintainer is open to it.
